pipeline/plugins/includes/tasks.py

```diff
@@ -0,0 +1,177 @@
+                                utils.binding_download_location(binding_id),
+                            )
+                            tmp_binding_path = os.path.join(
+                                tmpdir,
+                                utils.binding_download_location(binding_id),
+                            )
+
+                            SaveMetsSFTPOperator(
+                                task_id=f"save_mets_{binding_id}",
+                                api=api,
+                                sftp_client=sftp_client,
+                                ssh_client=ssh_client,
+                                tmpdir=tmp_binding_path,
+                                dc_identifier=dc_identifier,
+                                binding_path=binding_path,
+                                file_dir="mets",
+                            ).execute(context={})
+
+                            SaveAltosSFTPOperator(
+                                task_id=f"save_altos_{binding_id}",
+                                mets_path=f"{binding_path}/mets",
+                                sftp_client=sftp_client,
+                                ssh_client=ssh_client,
+                                tmpdir=tmp_binding_path,
+                                dc_identifier=dc_identifier,
+                                binding_path=binding_path,
+                                file_dir="alto",
+                            ).execute(context={})
+
+                            ssh_client.exec_command(f"rm -r {tmp_binding_path}")
+
+                @task(task_id=f"create_image_{set_id}_{image['prefix']}")
+                def create_image(image):
+                    print(f"Creating image_{set_id}_{image['prefix']} on Puhti")
+                    image_base_name = f"{set_id}_{image['prefix']}".rstrip("_")
+                    image_dir_path = os.path.join(base_path, image_base_name)
+
+                    ssh_hook = SSHHook(ssh_conn_id=ssh_conn_id)
+                    with ssh_hook.get_conn() as ssh_client:
+                        _, stdout, stderr = ssh_client.exec_command(
+                            f"mksquashfs {image_dir_path} {image_dir_path}.sqfs"
+                        )
+                        if stdout.channel.recv_exit_status() != 0:
+                            raise Exception(
+                                f"Creation of image {image_dir_path}.sqfs failed: {stderr.read().decode('utf-8')}"
+                            )
+                        ssh_client.exec_command(f"rm -r {image_dir_path}")
+
+                batch_downloads = []
+                for batch in utils.split_into_download_batches(image["bindings"]):
+                    batch_downloads.append(download_binding_batch(batch=batch))
+
+                prepare_download_location(image) >> batch_downloads >> create_image(image)
+
+            image_download_tg = download_image(image)
+            if image_downloads:
+                image_downloads[-1] >> image_download_tg
```

Comment:
`image_downloads` is appended to after this conditional. So, if `image_downloads` is still empty (aka during the creation of the first task group), a relation from the previous `image_download_tg` cannot be created because it doesn't exist. When there's at least one `image_download_tg` in the list, the following ones are are set to be run after the last one on the list.

Reply:
Ah, I already forgot that we are in a loop.